openstudiocore/src/model/ZonePropertyUserViewFactorsBySurfaceName.cpp

```diff
@@ -0,0 +1,505 @@
+    return addViewFactor(viewFactor);
+  }
+
+  bool ZonePropertyUserViewFactorsBySurfaceName_Impl::addViewFactor(const SubSurface& fromSubSurface, const SubSurface& toSubSurface, double value) {
+    ViewFactor viewFactor(fromSubSurface, toSubSurface, value);
+    return addViewFactor(viewFactor);
+  }
+
+  bool ZonePropertyUserViewFactorsBySurfaceName_Impl::addViewFactor(const SubSurface& fromSubSurface, const Surface& toSurface, double value) {
+    ViewFactor viewFactor(fromSubSurface, toSurface, value);
+    return addViewFactor(viewFactor);
+  }
+
+  bool ZonePropertyUserViewFactorsBySurfaceName_Impl::addViewFactor(const SubSurface& fromSubSurface, const InternalMass& toInternalMass, double value) {
+    ViewFactor viewFactor(fromSubSurface, toInternalMass, value);
+    return addViewFactor(viewFactor);
+  }
+
+  bool ZonePropertyUserViewFactorsBySurfaceName_Impl::addViewFactor(const InternalMass& fromInternalMass, const InternalMass& toSubSurface, double value) {
+    ViewFactor viewFactor(fromInternalMass, toSubSurface, value);
+    return addViewFactor(viewFactor);
+  }
+
+  bool ZonePropertyUserViewFactorsBySurfaceName_Impl::addViewFactor(const InternalMass& fromInternalMass, const Surface& toSurface, double value) {
+    ViewFactor viewFactor(fromInternalMass, toSurface, value);
+    return addViewFactor(viewFactor);
+  }
+
+  bool ZonePropertyUserViewFactorsBySurfaceName_Impl::addViewFactor(const InternalMass& fromInternalMass, const SubSurface& toSubSurface, double value) {
+    ViewFactor viewFactor(fromInternalMass, toSubSurface, value);
+    return addViewFactor(viewFactor);
+  }
+
+  bool ZonePropertyUserViewFactorsBySurfaceName_Impl::removeViewFactor(unsigned groupIndex) {
+    bool result = false;
+
+    unsigned int num = numberofViewFactors();
+    if (groupIndex < num) {
+      getObject<ModelObject>().eraseExtensibleGroup(groupIndex);
+      result = true;
+    }
+    return result;
+  }
+
+
+  void ZonePropertyUserViewFactorsBySurfaceName_Impl::removeAllViewFactors() {
+    getObject<ModelObject>().clearExtensibleGroups();
+  }
+
+  std::vector<ViewFactor> ZonePropertyUserViewFactorsBySurfaceName_Impl::viewFactors() const {
+    std::vector<ViewFactor> result;
+
+    for (unsigned i = 0; i < numberofViewFactors(); ++i) {
+
+      boost::optional<ViewFactor> _viewFactor = getViewFactor(i);
+
+      // getViewFactor is responsible for handling error and issuing Error log messages.
+      // Here we add it to the result array if it worked, and if it didn't, we keep going
+      // We just issue a message about index so user can delete it easily
+      if(_viewFactor) {
+        result.push_back(_viewFactor.get());
+      } else {
+        LOG(Error, briefDescription() << " has an invalid ViewFactor group at index " << i);
+      }
+    }
+
+    return result;
+  }
+
+  bool ZonePropertyUserViewFactorsBySurfaceName_Impl::addViewFactors(const std::vector<ViewFactor> &viewFactors) {
```

Comment:
So this method would return false if at least one view factor is not added (but the rest are added)? Or maybe the "Continuing with others." is throwing me- it's not actually continuing to add the remaining view factors, right?

Reply:
This is a choice really, but I thought if someone goes wrong you might want to return `false` in addition to logging a warn (or error) (as in "make sure you got what you want out of this method")